### PR TITLE
Modernizes TGUI

### PR DESCRIPTION
**Pull request by Sinestia**
## About The Pull Request

ports https://github.com/tgstation/tgstation/pull/54567 
https://github.com/tgstation/tgstation/pull/56080
https://github.com/tgstation/tgstation/pull/56225

edit: additional followup PRs will go here
https://github.com/tgstation/tgstation/pull/56223
https://github.com/tgstation/tgstation/pull/56713
https://github.com/tgstation/tgstation/pull/56727/
https://github.com/tgstation/tgstation/pull/56797
https://github.com/tgstation/tgstation/pull/57081
https://github.com/tgstation/tgstation/pull/57154
https://github.com/tgstation/tgstation/pull/57251
https://github.com/tgstation/tgstation/pull/56269
https://github.com/tgstation/tgstation/pull/57326
https://github.com/tgstation/tgstation/pull/57432
https://github.com/tgstation/tgstation/pull/57465
https://github.com/tgstation/tgstation/pull/57499
https://github.com/tgstation/tgstation/pull/57717
https://github.com/tgstation/tgstation/pull/57746
https://github.com/tgstation/tgstation/pull/57825
https://github.com/tgstation/tgstation/pull/57834
https://github.com/tgstation/tgstation/pull/57905
https://github.com/tgstation/tgstation/pull/58081
https://github.com/tgstation/tgstation/pull/58215
https://github.com/tgstation/tgstation/pull/58219
https://github.com/tgstation/tgstation/pull/58484
https://github.com/tgstation/tgstation/pull/58701
https://github.com/tgstation/tgstation/pull/56190

##  TGUI Changes for Coders
Stack component replaces <Flex spacing={1}>flex and generally replaces it, read more [here](https://github.com/tgstation/tgstation/pull/56080)
Storybook.js has been introduced, described in the above link. 

bin/tgui ran with flags is more robust at handling issues for you, remember to run --lint and --fix when submitting something (because I sure as hell kept forgetting and it makes you look bad anyways.)

Fragment without a key is replaced with <>.

Typescript 4.2 can now be utilized for tgui components. Unit tests can be ran using Jest. Described [here](https://github.com/tgstation/tgstation/pull/57081)

Custom SVG fonts can be utilized now! Just throw them in the tgui icon's directly and then you can utilize them in your scripts. This is described [here](https://github.com/tgstation/tgstation/pull/57717)

A prettierrc exists now, for those who care. Along with multiple linter and vscode changes in general. 

Resizable as a window property is dead, all windows are resizable by default now. 

**Some people have issues where their build differs from what the repo generates in the CIs, to fix this I believe you need to clear your untrackeds
you can do this with git clean -xdf on your tgui folder.**

**edit: on further investigation, it looks like some files have CRLF line endings instead of LF which they really shouldn't. Pulled straight from the repo, they're LF. This is the best lead I can find honestly.  Ping me Ithiria#9193 if there's issues or to confirm that any of this works or something.**

## Why It's Good For The Game

~~No unnecessary merge conflicts with tgui PRs. ~~ lol use a bot 

Newer TGUI

Typescript

Custom icons

Further monopolization of our lord and saviour vscode. 

## Changelog
:cl: Stylemistake, Cyberboss, SpaceManiac, other TG coders, Froststahr port. 
code: TGUI has been upgraded to 4.3 and now includes typescript. 
code: Custom SVG icons are now easy to do as well. 
fix: All tgui windows will resize automatically, and can now be resized as will. 
server: TGS will require adaptation. 
/:cl:

### CODE REVIEW

**Comment by ike709** [comment]
There's definitely more follow-up PRs.

**Comment by Sinestia** [comment]
oh. Fixing git tasks was that easy

**Comment by Sinestia** [comment]
> There's definitely more follow-up PRs.

Yep, will get them

Only thing I think I will leave out is tgui 4.3, it's big enough that I think it should be done separately.

**Comment by ike709** [comment]
If there's CBT changes in the 4.3 PR then get those but otherwise yes you can ignore 4.3.

**Comment by Sinestia** [comment]
> If there's CBT changes in the 4.3 PR then get those but otherwise yes you can ignore 4.3.

Seems like there's some important stuff in there, will have to get them then

edit: I Just went ahead and ported the whole thing, easier to do it all while I have the PR open instead of doing half
edit2: there's another PR I see, HmMm. Ill figure it later

**Comment by ike709** [comment]
https://github.com/BeeStation/BeeStation-Hornet/pull/4220

**Comment by Sinestia** [comment]
Was just gonna do the whole thing because that other PR had no documentation of what it actually ported but it's being reviewed so I guess I'll just not. 


 I think it's still bizarre since tg does these changes together (hence the CI screaming despite having the same change that "triggered it" lol) but w.e

**Comment by Sinestia** [comment]
image
Even modfab works as it used to 👌🏻

**Comment by Archanial** [comment]
epic :+1:

**Comment by ike709** [comment]
Does paper work? If this includes fully functional tgui 4.3 I'll go with this PR over the other one.

**Comment by Sinestia** [comment]
> Does paper work? If this includes fully functional tgui 4.3 I'll go with this PR over the other one.

Can't see anything obviously wrong about paper. 
image

File: .github/workflows/extra_pr_labels.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Check for conflicting PRs
-        uses: eps1lon/actions-label-merge-conflict@releases/2.x
+        uses: eps1lon/actions-label-merge-conflict@513a24fc7dca40990863be2935e059e650728400
```
Comment (Sinestia) [inline .github/workflows/extra_pr_labels.yml:14]: I couldn't figure out why, but bumping the version was necessary and there were no newer releases to use. TG was using a wholly different script.

**Comment by ike709** [comment]
Have you tested this locally? TGS refuses to testmerge this saying that it causes an error in DreamMaker.

**Comment by Sinestia** [comment]
> Have you tested this locally? TGS refuses to testmerge this saying that it causes an error in DreamMaker.

Yep, tested it locally quite a bit on both windows and linux to make sure everything builds fine and tguis work right.


It might not just work with TGS without other changes, to be frank I've never used TGS so I probably have neglected a necessary change to make it work. 

 I'll probably have to DL it to see what I've gotta do. 

edit: https://github.com/BeeStation/byond-docker 
this I presume


Edit2: `Good luck downstreams, remember to add the precompile script to your TGS instances.`
Weird, I do seem to have included the required scripts. I guess I will just have to download TGS and Figure It Out

**Comment by Sinestia** [comment]
image

Alright, to get it to work you must add PreCompile.bat to your EventScripts in TGS
(or PreCompile.sh for Linux.)

 The .yml should do it automatically for newly created instances. 

(The /tg/ version also created static folders for config and data, but I took that out. Can be added back in if it's wanted.)

**Review by Crossedfall — CHANGES_REQUESTED** [review]
Can't dig into it further right now, but the dockerfile will not build. Issue seems to be related to nodejs.

```
 => ERROR [build 4/4] RUN env TG_BOOTSTRAP_NODE_LINUX=1 tools/build/build     && tools/deploy.sh /deploy  && rm /  0.6s
------
 > [build 4/4] RUN env TG_BOOTSTRAP_NODE_LINUX=1 tools/build/build     && tools/deploy.sh /deploy       && rm /deploy/*.dll:
#15 0.500 realpath: ../bootstrap/.cache/node-v12.20.0.tar.gz: No such file or directory
------
executor failed running [/bin/sh -c env TG_BOOTSTRAP_NODE_LINUX=1 tools/build/build     && tools/deploy.sh /deploy     && rm /deploy/*.dll]: exit code: 1
```

**Comment by Sinestia** [comment]
Curious, node seems fine, the exit code I got is for it not being able to locate the file meaning the paths are probably screwed up. 

edit: actually our errors are different. it Might work for you now? 
edit2: probably not. Node will work but extools doesn't, glib errors w/ the prebuilt .so, likely from using the redone dockerfile from tg. I'll get around to it tm

File: .vscode/tasks.json [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			"group": {
+				"kind": "build",
+				"isDefault": true
+			},
+			"dependsOn": "dm: reparse",
+			"label": "Build All"
+		},
+		{
+			"type": "dreammaker",
+			"dme": "tgstation.dme",
```
Comment (ike709) [inline .vscode/tasks.json]: 🤔

**Comment by ike709** [comment]
After checking this out locally and attempting to compile, it errors with `command 'dreammaker.reparse' not found`.

**Comment by Sinestia** [comment]
> 
> 
> After checking this out locally and attempting to compile, it errors with `command 'dreammaker.reparse' not found`.

I can't replicate this, I'm on a fresh install too.

**Comment by Sinestia** [comment]
Right, back to my setup, gonna try to resolve the docker situation 2morrow

line endings on files w/o extensions (so not in gitattributes) and compiling extools are the problems

**Comment by AffectedArc07** [comment]
You dont want to do this

The amount of comments I have seen in coderbus about this shit being broken and people whining that they cant easily build, not to mention CBT builds end up taking longer for people (I have no idea why, maybe from running a subprocesses out of node), but this is a step in the wrong direction. 

You have to remember that not everyone who downloads the code is a coder, and some people just want it for local test servers. These people just press the dme thats called `beestation` and manually do the build inside the DM "IDE". They dont want to have to be installing extra shit, or worse, having extra shit installed on their PC without them even realising.

**EDIT:** I literally just had to guide TGMC through making this shit work on TGS3.

**Comment by Sinestia** [comment]
Docker functions. I admittedly am clueless with docker and it was much easier to just change the old one mostly ignoring tg changes due to extools.... not playing well even when I made it compile itself.

**Comment by ike709** [comment]
Maintainers are currently considering if we want to use CBT or not, given the mixed opinions in coderbus.

**Comment by MCterra10** [comment]
The tgui upgrade is great, but I heavily disagree with forcing every coder who doesn't give a shit about tgui and doesn't work with it to install the tumour that is node on their system to be able to compile the game

**Comment by Sinestia** [comment]
NodeJS is soy software, though you don't really install it. 

It just keeps a local dir of it, it's a cyst, not a tumor.

**Comment by ike709** [comment]
The maintainers and downstream have decided that we don't want the overhead introduced by CBT. I'm willing to leave this PR open for the general TGUI improvements/updates if you remove the CBT part.

**Comment by ike709** [comment]
The topic will be revisited once Juke Build has matured.

**Comment by Sinestia** [comment]
CBT will be removed after Coderboy Arguments, other modernizations will remain

**Comment by ike709** [comment]
The arguments have been had, CBT isn't happening.

**Comment by Sinestia** [comment]
git will hate this not done yet but finally started getting around to it

**Comment by jupyterkat** [comment]
Code and Build Torture

**Comment by Sinestia** [comment]
Threw in one more to make reviewing easier and more exciting

https://github.com/tgstation/tgstation/pull/56727/

should be ready 2 be looked at again, and might definitely wanna TM lol. 

edit: pleas eoh god bee tgui breaks for me as a client on 514 but this doesnt ahahaE

**Comment by BrumalLake** [comment]
tell me this includes the tall ass atmos chamber monitoring consoles

**Comment by Sinestia** [comment]
> 
> 
> tell me this includes the tall ass atmos chamber monitoring consoles

 You can think of this as being solely about the tgui backend

not any specific interfaces were changed beyond updating them

**Comment by MCterra10** [comment]
Yolomerged

**Comment by MCterra10** [comment]
Yolo

**Comment by ike709** [comment]
It was testmerged on sage.

**Comment by jupyterkat** [comment]
ci checks fails on tgui regardless if the build is correct or not

**Comment by BrumalLake** [comment]
> 
> 
> ci checks fails on tgui regardless if the build is correct or not

https://github.com/BeeStation/BeeStation-Hornet/pull/4606
no it's somehow possible, not sure how i got it

**Comment by Sinestia** [comment]
> 
> 
> ci checks fails on tgui regardless if the build is correct or not

Your cache/untrackeds are bad I believe

run git clean -xdf on your tgui folder

happened to me too


edit: on further investigation, it looks like some files have CRLF line endings instead of LF which they really shouldn't. Pulled straight from the repo, they're LF. This is the only tangible difference between my folder and that of someones who didn't work.

**Pull request merged**